include/ilang/ila/var_container.h

```diff
@@ -0,0 +1,391 @@
+  typedef std::vector<std::pair<std::string, VarContainerPtr>> struct_container;
+  // [TODO] names are available from the type, can save a ton of space if I create a custom
+  // container that doesn't store string names.
+
+  // [TODO] consider implementing a partitions as a VarContainer
+  typedef std::vector<VarContainerPtr> partition;
+  /// Right now partitions are not a VarContainer Vector because partitions may
+  /// have different size.
+
+  // [NOTE] constructor is protected -- use Make instead.
+
+  /// Makes a VarContainer representing a complex object of the given type, 
+  /// using the given name as a prefix for the state/input variables needed to 
+  /// define the object.
+  static VarContainerPtr Make(const std::string& name, const SortPtr& sort);
+
+  /// Converts an Expr to the equivalent VarContainer Primitive
+  static VarContainerPtr from_primitive_expr(const ExprPtr& p);
+
+  /// Converts a similarly structured c++ object over ExprPtrs and VarContainerPtrs 
+  /// to a VarContainerPtr.
+  /// Use with extreme caution -- it's very easy to abuse, but may be better than
+  /// ad-hoc construction because it does some typechecking.
+  template<typename T>
+  static VarContainerPtr from_cpp_obj(const SortPtr& sort, std::vector<T> obj);
+
+  /// Converts a similarly structured c++ object over ExprPtrs and VarContainerPtrs 
+  /// to a VarContainerPtr.
+  /// Use with extreme caution -- it's very easy to abuse, but may be better than
+  /// ad-hoc construction because it does some typechecking.
+  template<typename It>
+  static VarContainerPtr from_cpp_obj(const SortPtr& sort, It struct_begin, It struct_end);
+
+  /// Converts a similarly structured c++ object over ExprPtrs and VarContainerPtrs 
+  /// to a VarContainerPtr.
+  /// Use with extreme caution -- it's very easy to abuse, but may be better than
+  /// ad-hoc construction because it does some typechecking.
+  template<typename T>
+  static VarContainerPtr from_cpp_obj(const SortPtr& sort, std::vector<std::pair<std::string, T>> obj) {
+    return from_cpp_obj(sort, obj.begin(), obj.end());
+  }
+
+  /// Converts a similarly structured c++ object over ExprPtrs and VarContainerPtrs 
+  /// to a VarContainerPtr.
+  /// Use with extreme caution -- it's very easy to abuse, but may be better than
+  /// ad-hoc construction because it does some typechecking.
+  template<typename T>
+  static VarContainerPtr from_cpp_obj(const SortPtr& sort, std::unordered_map<std::string, T> obj) {
+    return from_cpp_obj(sort, obj.begin(), obj.end());
+  }
+
+  /// Converts a similarly structured c++ object over ExprPtrs and VarContainerPtrs 
+  /// to a VarContainerPtr.
+  /// Use with extreme caution -- it's very easy to abuse, but may be better than
+  /// ad-hoc construction because it does some typechecking.
+  static inline VarContainerPtr from_cpp_obj(const SortPtr& sort, VarContainerPtr obj);
```

Comment:
This is actually using function overloading to handle a base case of `from_cpp_obj`. Note that `from_cpp_obj` is also a templated function that recurses down on its childen. If those children are SortPtrs, then this function will be called instead of the templated function.